utils/src/activityLog/Activity.ts

```diff
@@ -0,0 +1,87 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import { randomUUID } from "crypto";
+import { EventEmitter, TreeItemCollapsibleState } from "vscode";
+import * as types from '../../index';
+import { parseError } from "../parseError";
+import { AzExtTreeItem } from "../tree/AzExtTreeItem"
+
+export interface ActivityTreeItemOptions {
+    label: string;
+    contextValuePostfix?: string;
+    collapsibleState?: TreeItemCollapsibleState;
+    children?: (parent: types.AzExtParentTreeItem) => AzExtTreeItem[];
+}
+
+export abstract class ActivityBase implements types.Activity {
+
+    public readonly onStart: typeof this._onStartEmitter.event;
+    public readonly onProgress: typeof this._onProgressEmitter.event;
+    public readonly onSuccess: typeof this._onSuccessEmitter.event;
+    public readonly onError: typeof this._onErrorEmitter.event;
+
+    private readonly _onStartEmitter: EventEmitter<types.OnStartActivityData>;
+    private readonly _onProgressEmitter: EventEmitter<types.OnProgressActivityData>;
+    private readonly _onSuccessEmitter: EventEmitter<types.OnSuccessActivityData>;
+    private readonly _onErrorEmitter: EventEmitter<types.OnErrorActivityData>;
+
+    public running: boolean;
```

Comment:
To force data consistency, the state should be tracked as one enum, (e.g. `'NotStarted' | 'Running' | 'Success' | 'Failed' | 'Cancelled'`), rather than two booleans.